blockchain/indexbuilder.go

```diff
@@ -113,12 +119,145 @@ func (ib *IndexBuilder) HandleBlock(blk *block.Block) error {
 	ib.pendingBlks <- blk
 	return nil
 }
+func initIndexActionsKey(store db.KVStore) error {
+	_, err := store.Get(blockActionBlockMappingNS, indexActionsTipIndexKey)
+	if err != nil && errors.Cause(err) == db.ErrNotExist {
+		if err = store.Put(blockActionBlockMappingNS, indexActionsTipIndexKey, make([]byte, 8)); err != nil {
+			return errors.Wrap(err, "failed to write initial value for tip index of index actions")
+		}
+	}
 
+	_, err = store.Get(blockActionBlockMappingNS, indexActionsTipHeightKey)
+	if err != nil && errors.Cause(err) == db.ErrNotExist {
+		if err = store.Put(blockActionBlockMappingNS, indexActionsTipHeightKey, make([]byte, 8)); err != nil {
+			return errors.Wrap(err, "failed to write initial value for tip height of index actions")
+		}
+	}
+	return err
+}
+func (ib *IndexBuilder) getStartHeightAndIndex() (startHeight, startIndex uint64, err error) {
+	// get index that already builded
+	startIndex, err = getNextIndex(ib.store)
+	if err != nil {
+		return
+	}
+	// get height that already builded
+	startHeight, err = getNextHeight(ib.store)
+	if err != nil {
+		return
+	}
+	return
+}
+func (ib *IndexBuilder) commitBatchAndClear(tipIndex, tipHeight uint64, batch db.KVStoreBatch) error {
+	tipIndexBytes := byteutil.Uint64ToBytes(tipIndex)
+	batch.Put(blockActionBlockMappingNS, indexActionsTipIndexKey, tipIndexBytes, "failed to put tip index of actions")
+	tipHeightBytes := byteutil.Uint64ToBytes(tipHeight)
+	batch.Put(blockActionBlockMappingNS, indexActionsTipHeightKey, tipHeightBytes, "failed to put tip height")
+	if err := ib.store.Commit(batch); err != nil {
```

Comment:
Is it safe to reuse the batch without cleanup the data inside? Mutable way to use a data struct is more easily to introduce bugs.

Reply:
commit method will clean batch if commit success